### PR TITLE
Add ADR  on cloud.gov

### PR DESCRIPTION
**Pull request by rpdelaney**
# [OHSH-229](https://ocio-jira.acf.hhs.gov/browse/OHSH-229)

Changes proposed in this pull request:

- Add first draft of proposed ADR : "On whether to migrate hosted OHS apps to cloud.gov or not"
- Ignore markdownlint "ul-style" rule since it was fighting with markdownlint-toc hook

I left `?`s where I feel there may be gaps others could fill in.

### CODE REVIEW

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  * [Migrate everything to cloud.gov all at once](#migrate-everything-to-cloudgov-all-at-once)
+  * [Migrate one app to cloud.gov](#migrate-one-app-to-cloudgov)
+  * [Wait until the hosted apps are ready for cloud.gov, then migrate everything](#wait-until-the-hosted-apps-are-ready-for-cloudgov-then-migrate-everything)
+
+<!-- Regenerate with "pre-commit run -a markdown-toc" -->
+
+<!-- tocstop -->
+
+## Context and Problem Statement
+
```
Comment (ssatagopan) [inline docs/adr/0011-on-migration-to-cloud.gov.md:30]: Some additional problems/context:
* Infrastructure and applications aren't built with horizontal scalability in mind (leading sometimes to responsiveness issues)
* App teams aren't able to confidently say that their applications will work in production environment because of differences in current infrastructure between dev/staging and prod

Found some earlier conversations on Slack also: https://trussworks.slack.com/archives/C01T6F5PB1A/p1631909491116900
https://trussworks.slack.com/archives/C01T6F5PB1A/p1632147056124200
https://trussworks.slack.com/archives/C01T6F5PB1A/p1633637098189500

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+- ?
+
+## Decision Drivers
+
+- We want to reduce the dependency of the app teams on the hosting team, to give
+  them autonomy and to allow us to focus on our broad strategy goals.
+- App teams need access to infrastructure on an as-needed basis consistent with
+  the principle of least privilege.
+- There is a substantial flat cost for using cloud.gov.
+- The project has a limited bugdet for infrastructure costs.
```
Comment (ssatagopan) [inline docs/adr/0011-on-migration-to-cloud.gov.md:52]:
```suggestion
- The project has a limited budget for infrastructure costs.
```

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+- We need to maintain the stability of the applications in production during
+  infrastructure level transitions.
+
+## Considered Options
+
+- Build our own self-managed, comprehensive solution in AWS
+- Migrate everything to cloud.gov all at once
+- Migrate one app to cloud.gov
+- Wait until the hosted apps are ready for cloud.gov, then migrate everything
```
Comment (ssatagopan) [inline docs/adr/0011-on-migration-to-cloud.gov.md:61]: I know this is based on fairly recent conversations with Dan, but could we modify this bullet to say:
"Shore up existing infrastructure and help apps become 12F enough so they are ready for cloud.gov or any other platform"

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+## Decision Outcome
+
+Wait until the hosted apps are ready for cloud.gov, then migrate everything, because
```
Comment (ssatagopan) [inline docs/adr/0011-on-migration-to-cloud.gov.md:65]: Would rephrase as per my previous comment that option is "Shoring up existing infrastructure and help apps become 12F enough..."

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+this is the only option that meets decision drivers around transition stability and
+budget constraints.
+
+### Positive Consequences
+
+- We have a clear milestone to work toward that motivates needed modernization
```
Comment (ssatagopan) [inline docs/adr/0011-on-migration-to-cloud.gov.md:71]: Also lets us improve current infrastructure to make it more sustainable

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  unacceptable risk of a major unplanned outage.
+
+### Migrate one app to cloud.gov
+
+- Good, because we can test cloud.gov with a smaller slice of the architecture
+  as a proof of concept.
+- Bad, because there is a substantial flat cost for using cloud.gov, and we have decided that
+  having only one app there is not a worthwhile use of the current budget.
+
+### Wait until the hosted apps are ready for cloud.gov, then migrate everything
```
Comment (ssatagopan) [inline docs/adr/0011-on-migration-to-cloud.gov.md:109]: Same comment as above about adjusting phrasing to reflect 12F enough...

**Review by ssatagopan — CHANGES_REQUESTED** [review]

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+- Good, because we can design a made-to-order solution according to our specific
+  requirements.
+- Good, because we have direct control over how all the AWS services are integrated.
+- Good, because we have access to more AWS services in general.
+- Bad, because we have inadequate engineering bandwidth to put everything together
+  and get improvements rolled out to our users within an acceptable time frame.
+
+### Migrate everything to cloud.gov all at once
+
+- Good, because this is the shortest path to solving all the problems.
```
[the remark this answers is not in the document]
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md:94]: How about this?

```suggestion
- Good, because this is the shortest path to solving all the problems described in the context and problem statement.
```

File: docs/adr/0011-on-migration-to-cloud.gov.md (hunk 1 above)
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: > Infrastructure and applications aren't built with horizontal scalability in mind (leading sometimes to responsiveness issues)

Just checking - is this something that was specifically considered in the cloud.gov decision? ("Yes" is plenty good enough as an answer. just making sure I'm following)

> App teams aren't able to confidently say that their applications will work in production environment because of differences in current infrastructure between dev/staging and prod

Ah yes, Rebecca mentioned this also but I lost it in my disorganized notes. I added something about that point.

File: docs/adr/0011-on-migration-to-cloud.gov.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: 3f0e75b

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+## Pros and Cons of the Options
+
+### Build our own self-managed, comprehensive solution in AWS
+
+- Good, because we can design a made-to-order solution according to our specific
+  requirements.
+- Good, because we have direct control over how all the AWS services are integrated.
+- Good, because we have access to more AWS services in general.
+- Bad, because we have inadequate engineering bandwidth to put everything together
```
[the remark this answers is not in the document]
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: 8a47ebc

File: docs/adr/0011-on-migration-to-cloud.gov.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  and get improvements rolled out to our users within an acceptable time frame.
+
+### Migrate everything to cloud.gov all at once
+
+- Good, because this is the shortest path to solving all the problems.
+- Bad, because cloud.gov requires containerization, and several hosted apps are currently
```
[the remark this answers is not in the document]
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: Does this get it? 3464dae

File: docs/adr/0011-on-migration-to-cloud.gov.md (hunk 5 above)
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: What does 'sustainable' mean?

File: docs/adr/0011-on-migration-to-cloud.gov.md (hunk 1 above)
Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: I'm concerned about constraining the scope of the ADR to what was actually considered a decision driver connected to the problem statement when the decision was being made. Can we chat about this offline?

File: docs/adr/0011-on-migration-to-cloud.gov.md (hunk 5 above)
Comment (ssatagopan) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: > What does 'sustainable' mean?

Easier for hosting team to manage (less manual work, confidence in being resilient to user demand and resistant to malicious activity), more self-service for app teams

Comment (rpdelaney) [reply docs/adr/0011-on-migration-to-cloud.gov.md]: c9fd9c0

**Review by reid-truss — COMMENTED** [review]
I think a couple of Sanjay's comments might still be outstanding, but other than that, LGTM - thanks for researching and capturing this.

File: docs/adr/0012-on-migration-to-cloud.gov.md [hunk 10]
```diff
@@ -105,13 +112,18 @@ budget constraints.
 
 - Good, because we can test cloud.gov with a smaller slice of the architecture
   as a proof of concept.
+- Good, because our [research][1] suggests that we have at least one app (Coaching Companion)
+  that is close to ready for migration.
 - Bad, because there is a substantial flat cost for using cloud.gov, and we have decided that
   having only one app there is not a worthwhile use of the current budget.
```
[the remark this answers is not in the document]
Comment (ssatagopan) [reply docs/adr/0012-on-migration-to-cloud.gov.md:119]: This suggestion is for the opportunity cost of migrating one app vs. doing the shoring up work and getting to dev ops work

**Review by ssatagopan — APPROVED** [review]
